hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/ReliableTopicProxy.java

```diff
@@ -255,4 +276,138 @@ protected void postDestroy() {
+
+        try {
+            List<ReliableTopicMessage> messages = payload.stream()
+                    .map(m -> new ReliableTopicMessage(nodeEngine.toData(m), thisAddress))
+                    .collect(Collectors.toList());
+            switch (overloadPolicy) {
+                case ERROR:
+                    long sequenceId = ringbuffer.addAllAsync(messages, OverflowPolicy.FAIL).toCompletableFuture().get();
+                    if (sequenceId == -1) {
+                        throw new TopicOverloadException(
+                                String.format("Failed to publish messages: %s on topic: %s", payload, getName()));
+                    }
+                    break;
+                case DISCARD_OLDEST:
+                    ringbuffer.addAllAsync(messages, OverflowPolicy.OVERWRITE).toCompletableFuture().get();
+                    break;
+                case DISCARD_NEWEST:
+                    ringbuffer.addAllAsync(messages, OverflowPolicy.FAIL).toCompletableFuture().get();
+                    break;
+                case BLOCK:
+                    addWithBackoff(messages);
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unknown overloadPolicy:" + overloadPolicy);
+            }
+
+            localTopicStats.incrementPublishes();
+        } catch (Exception e) {
+            throw (RuntimeException) peel(e, null,
+                    String.format("Failed to publish messages: %s on topic: %s", payload, getName()));
+        }
+    }
+
+    @Override
+    public CompletionStage<Void> publishAllAsync(@Nonnull Collection<? extends E> payload) {
+        checkNotNull(payload, NULL_MESSAGE_IS_NOT_ALLOWED);
+        checkNoNullInside(payload, NULL_MESSAGE_IS_NOT_ALLOWED);
+
+        // RU_COMPAT_4_0
+        checkClusterVersion(Versions.V4_1);
+
+        InternalCompletableFuture<Void> returnFuture = new InternalCompletableFuture<>();
+        try {
+            List<ReliableTopicMessage> messages = payload.stream()
+                    .map(m -> new ReliableTopicMessage(nodeEngine.toData(m), thisAddress))
+                    .collect(Collectors.toList());
+            switch (overloadPolicy) {
+                case ERROR:
+                    addAsyncOrFail(payload, returnFuture, messages);
+                    break;
+                case DISCARD_OLDEST:
+                    addAsync(messages, OverflowPolicy.OVERWRITE);
+                    break;
+                case DISCARD_NEWEST:
+                    addAsync(messages, OverflowPolicy.FAIL);
+                    break;
+                case BLOCK:
+                    addAsyncAndBlock(payload, returnFuture, messages, INITIAL_BACKOFF_MS);
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unknown overloadPolicy:" + overloadPolicy);
+            }
+        } catch (Exception e) {
+            returnFuture.completeExceptionally(peel(e, null,
```

Comment:
In the member-side we complete the future in the client-side we throw directly. We should probably have the same behaviour and between the two, I think I'd prefer the client-side one (throwing directly) as it's an early indicator of failure, like the instance already been shutdown, or the message not being serializable.

Reply:
Fixed